### PR TITLE
Remote Repository and  Remote Organization Normalization

### LINKED ISSUE #7572: repository import filter doesn't
## Details

I'm trying to add a project (cjw296/giterator), so selected just cjw296

Screenshot 2020-10-19 at 08 23 43

## Expected Result

Only cj296 projects to be shown.

## Actual Result

A mish-mash of projects.

As a feature request related to this, it would be great to be able to search for a repo on this screen. conda-forge is impossible to navigate by paging through...

### LINKED ISSUE #7727: Race condition with get_or_create in remote repositories
Since we don't have `unique_together` in the model, and we use get_or_create there is race condition where users end up with duplicates, giving an error like "get() returned 2 remote repositories"

https://docs.djangoproject.com/en/2.2/ref/models/querysets/#get-or-create

https://github.com/readthedocs/readthedocs.org/blob/811890894f72c14be05ddd443a20adc017656c5c/readthedocs/oauth/services/bitbucket.py#L126-L129

`account` and `full_name` should be unique together. Looks like this would be fixed with the remote repository refactor.

https://github.com/orgs/readthedocs/projects/74

Current work around is to manually delete the duplicates and trigger a resync.

### PR DESCRIPTION
**Pull request by saadmk11**
This PR contains the full implementation for the Remote Repository and  Remote Organization Normalization.

This PR creates separate database tables for the `RemoteRepository` and `RemoteOrganization` with the new modeling.
This also updates the codebase with the changes for the new modeling and adds a management command to re-sync all the remote repositories and remoter organizations.

The tasks done in this PR can be found here: https://github.com/orgs/readthedocs/projects/74#column-11164063

Closes #7572 
Closes #7727

### CODE REVIEW

**Comment by humitos** [comment]
It'd be good to get another look from someone from the @readthedocs/core team to double check that we are not missing anything here.

Following steps are:

- create a packer image with this code _without merging it into master yet_
- spin up an instance with this code
- run the `migrate` command from this instance
- run management command `sync_vcs_data` for our own users
- open site from this instance and check that we do see our repositories from "Import Project" list

After this code is merged into `master` we should dump the `Project -> RemoteRepository` relations from the old instance and load them into the new instance at deploy while doing the switch.

> Marking as `blocked` to avoid merging by mistake for now.

File: readthedocs/api/v2/serializers.py
Comment (ericholscher) [inline readthedocs/api/v2/serializers.py:205]: Where is this cache set?

File: readthedocs/api/v2/views/model_views.py
Comment (ericholscher) [inline readthedocs/api/v2/views/model_views.py:332]: Do we need this query if we default to `False`? Seems we just need to check if they're an admin?

File: readthedocs/api/v2/views/model_views.py
Comment (ericholscher) [inline readthedocs/api/v2/views/model_views.py:348]: What is `own` here? Could probably use a better name

File: readthedocs/core/signals.py
Comment (ericholscher) [inline readthedocs/core/signals.py:107]: Why did we remove this code? Do we run it somewhere else?

File: readthedocs/oauth/models.py
Comment (ericholscher) [inline readthedocs/oauth/models.py:61]: This syntax feels weird. Sounds like we can just do a single list: https://docs.djangoproject.com/en/2.2/ref/models/options/#django.db.models.Options.unique_together

```suggestion
        unique_together = ('remote_id', 'vcs_provider')
```

File: readthedocs/oauth/models.py
Comment (ericholscher) [inline readthedocs/oauth/models.py:190]: I wonder if having a FK lookup in the ordering query here makes it slower. Do we always want this, or only when doing the import form? If only on import, we should probably just force ordering there.

File: readthedocs/oauth/services/base.py
Comment (ericholscher) [inline readthedocs/oauth/services/base.py:231]: Should this be on the model, instead of a util here?

File: readthedocs/oauth/services/bitbucket.py
Comment (ericholscher) [inline readthedocs/oauth/services/bitbucket.py:136]: We should probably document somewhere that the `remote_id` is for each VCS provider.

File: readthedocs/oauth/services/bitbucket.py
Comment (ericholscher) [inline readthedocs/oauth/services/bitbucket.py:200]: It feels kind of magical that `self.account` is implicitly set on the relation in this function. I don't know if it's _bad_, but it feels a bit weird reading this code, and wondering how the account gets set on the relation.

File: readthedocs/oauth/services/github.py
Comment (ericholscher) [inline readthedocs/oauth/services/github.py:167]: This seems like a weird mapping. Could probably use a comment.

**Review by ericholscher — APPROVED** [review]
Looks good for testing in prod. We should figure out the DB constraint issue tho 👍

File: readthedocs/api/v2/serializers.py
Comment (saadmk11) [reply readthedocs/api/v2/serializers.py:205]: We discussed about this here https://github.com/readthedocs/readthedocs.org/pull/7778#discussion_r549993292

File: readthedocs/api/v2/views/model_views.py
Comment (saadmk11) [reply readthedocs/api/v2/views/model_views.py:332]: Yup, we can just remove it. :)

File: readthedocs/api/v2/views/model_views.py
Comment (saadmk11) [reply readthedocs/api/v2/views/model_views.py:348]: I'm not sure why it's called `own` here, I think it refers to the VCS provider names ('github', 'gitlab' etc). We can change the variable name but the quey param name is for the API V2

File: readthedocs/core/signals.py
Comment (saadmk11) [reply readthedocs/core/signals.py:107]: We are not deleting `RemoteOrganization` on user delete as we will only delete the `RemoteOrganizationRelation` in this case and that is handled by model `on_delete=CASECADE`.

File: readthedocs/oauth/services/base.py
Comment (saadmk11) [reply readthedocs/oauth/services/base.py:231]: We are only using this here and we already have access to the account and user object here as well, that's why felt like it should be here so that we don't need to pass these to other places, but we can probably move it to the models.

File: readthedocs/api/v2/views/model_views.py
Comment (humitos) [reply readthedocs/api/v2/views/model_views.py:343]: We could probably make a note on this for the APIv3 where these endpoints are not created yet: https://github.com/readthedocs/readthedocs.org/pull/7510

File: readthedocs/core/signals.py
Comment (humitos) [reply readthedocs/core/signals.py:107]: If we want, we could do a cleanup of organizations/repositories at some point by deleting those that has 0 users connected. I don't think we need this _now_, tho.

File: readthedocs/oauth/services/base.py
Comment (humitos) [reply readthedocs/oauth/services/base.py]: I also prefer to have the code contained here, in particular if it's not required outside, and also to avoid keep adding lot of model methods if they are not strictly required there.

Comment (humitos) [reply readthedocs/oauth/services/base.py]: Don't worry too much. I'm fine having it on this model as well. I thought it was ending on the `Project` model, but I was wrong.

File: readthedocs/oauth/models.py
Comment (humitos) [reply readthedocs/oauth/models.py:268]: I'd say that this is "not strictly required". However, for some actions in the old code we tried first with this account (the one the originally imported the repo) and if fail it continues with the following. This reduced the probabilities to fail due lack of permissions. (*)

In case we want to do this here, we will need to modify some code to keep this synced/updated. Besides, if we rely on `imported` for specific tasks we will need to write fallback code in case the original user deletes his account: we won't have the `RemoteRepositoryRelation.imported=True` anymore for that `Project`.

(*) we could use your idea but rename it to `prefered=True` (or similar) that we can set with the first account that works, and update with the following in the list when it fails. Makes sense?

**Comment by humitos** [comment]
I did some research and tests for the database constraint that we will need to deal with. The plan here will be:

1. remove all the constraints from the new tables (`oauth_remoterepository_2020`, `oauth_remoterepository_relations`, `oauth_remoteorganization_relations`) that reference to current used tables (`socialaccount_socialaccount`, `auth_user`, etc) that could be deleted while doing the 1-week resync
2. do the resync during that week
3. re-add these constraints as `NOT VALID` (more below) so deleted users/accounts are not validated, but all new data is validated before insert/update
4. remove all rows from new tables that are not valid (the row referenced in the other table was deleted) --this step is not strict, but is a good cleanup

PostgreSQL documentation: https://www.postgresql.org/docs/9.1/sql-altertable.html

> ADD table_constraint [ NOT VALID ]
>     This form adds a new constraint to a table using the same syntax as CREATE TABLE, plus the option NOT VALID, which is currently only allowed for foreign key constraints. If the constraint is marked NOT VALID, the potentially-lengthy initial check to verify that all rows in the table satisfy the constraint is skipped. The constraint will still be enforced against subsequent inserts or updates (that is, they'll fail unless there is a matching row in the referenced table). But the database will not assume that the constraint holds for all rows in the table, until it is validated by using the VALIDATE CONSTRAINT option.

As example in my local instance:

- DROP CONSTRAINT

```sql
# Drop constraints from RemoteRepositoryRelation
ALTER TABLE oauth_remoterepositoryrelation DROP CONSTRAINT oauth_remotereposito_account_id_02d56944_fk_socialacc;
ALTER TABLE oauth_remoterepositoryrelation DROP CONSTRAINT oauth_remoterepositoryrelation_user_id_3c2baf85_fk_auth_user_id;

# Drop constraints from RemoteOrganizationRelation
ALTER TABLE oauth_remoteorganizationrelation DROP CONSTRAINT oauth_remoteorganiza_account_id_8dbef288_fk_socialacc;
ALTER TABLE oauth_remoteorganizationrelation DROP CONSTRAINT oauth_remoteorganiza_user_id_0603492b_fk_auth_user;
```

- ADD CONSTRAINT
```sql

# Add constraints from RemoteRepositoryRelation
ALTER TABLE "oauth_remoterepositoryrelation" ADD CONSTRAINT "oauth_remoterepositoryrelation_user_id_3c2baf85_fk_auth_user_id" FOREIGN KEY ("user_id") REFERENCES "auth_user" ("id") DEFERRABLE INITIALLY DEFERRED;

ALTER TABLE "oauth_remoterepositoryrelation" ADD CONSTRAINT "oauth_remotereposito_account_id_02d56944_fk_socialacc" FOREIGN KEY ("account_id") REFERENCES "socialaccount_socialaccount" ("id") DEFERRABLE INITIALLY DEFERRED;


# Add constraints for RemoteOrganizationRelation
ALTER TABLE "oauth_remoteorganizationrelation" ADD CONSTRAINT "oauth_remoteorganiza_user_id_0603492b_fk_auth_user" FOREIGN KEY ("user_id") REFERENCES "auth_user" ("id") DEFERRABLE INITIALLY DEFERRED;

ALTER TABLE "oauth_remoteorganizationrelation" ADD CONSTRAINT "oauth_remoteorganiza_account_id_8dbef288_fk_socialacc" FOREIGN KEY ("account_id") REFERENCES "socialaccount_socialaccount" ("id") DEFERRABLE INITIALLY DEFERRED;
```

First, try to `ADD CONTRAINT` as they are here (changing the name, probably for the production ones) and if they fail saying that there are some invalid data, add `NOT VALID` at the end. After that, do the cleanup step.

**Comment by ericholscher** [comment]
`NOT VALID` is awesome!

File: readthedocs/oauth/models.py
Comment (ericholscher) [reply readthedocs/oauth/models.py:268]: I don't think it's important to have then, unless we have a strong reason to have it. Seems like other prioritization methods (eg. owners then team members, etc.) would work fine, but also just "everyone who has access" is fine for now.

File: readthedocs/api/v2/serializers.py
Comment (ericholscher) [reply readthedocs/api/v2/serializers.py:205]: Ha, we should add a comment then, if we both had the same questions :)

**Review by ericholscher — APPROVED** [review]
Changes look good!

**Comment by humitos** [comment]
Just to mention that I linked docs from 9.1 and we are using v12. The documentation is a updated and better: https://www.postgresql.org/docs/12/sql-altertable.html

> First, try to `ADD CONSTRAINT` as they are here (changing the name, probably for the production ones) and if they fail saying that there are some invalid data, add `NOT VALID` at the end. After that, do the cleanup step.

After reading the v12 docs it seems it's better to use `NOT VALID` at first (to avoid a huge scan on the table that will lock it) and once it's done use `VALIDATE CONSTRAINT` that won't lock the table. However, from the Notes section,

> validation acquires only a SHARE UPDATE EXCLUSIVE lock on the table being altered. (If the constraint is a foreign key then a ROW SHARE lock is also required on the table referenced by the constraint.)

it seems we may need to lock the other tables? (Project and SocialAcctount)

**Comment by humitos** [comment]
This PR is merged in `rel` and already deployed in Community. I'm not merging it into `master` yet because we don't want to be deployed in Commercial yet. We are still doing some QA on .org and we haven't triggered the re-sync task on .com yet.

**Review by humitos — APPROVED** [review]
This PR is ready to be merged and it will go out tomorrow in commercial.

**Pull request merged**
